src/garage/_dtypes.py

```diff
@@ -167,3 +167,199 @@ def __new__(cls, env_spec, observations, actions, rewards, terminals,
+
+        if not env_spec.observation_space.contains(first_next_observation):
+            raise ValueError(
+                'next_observations must conform to observation_space {}, but got '
+                'data with shape {} instead.'.format(
+                    env_spec.observation_space, first_next_observation))
+
+        # actions
+        # TODO: Shape check can be removed for gym>=0.15.4 # pylint: disable=fixme # noqa: E501
+        if not (env_spec.action_space.contains(first_action)
+                and first_action.shape == env_spec.action_space.shape):
+            raise ValueError(
+                'actions must conform to action_space {}, but got data with '
+                'shape {} instead.'.format(env_spec.action_space,
+                                           first_action))
+
+        if terminals.dtype != np.bool:
+            raise ValueError(
+                'Terminals tensor must be dtype np.bool, but got tensor '
+                'of dtype {} instead.'.format(terminals.dtype))
+
+        # env_infos
+        for key, val in env_infos.items():
+            if not isinstance(val, np.ndarray):
+                raise ValueError(
+                    'Each entry in env_infos must be a numpy array, but got '
+                    'key {} with value type {} instead.'
+                    'instead'.format(key, type(val)))
+
+        # agent_infos
+        for key, val in agent_infos.items():
+            if not isinstance(val, np.ndarray):
+                raise ValueError(
+                    'Each entry in agent_infos must be a numpy array, but got '
+                    'key {} with value type {} instead.'
+                    'instead'.format(key, type(val)))
+
+        return super().__new__(SamplesBatch, env_spec, observations,
+                               actions, rewards, next_observations, terminals,
+                               env_infos, agent_infos)
+
+
+class Sample(
+        collections.namedtuple('Sample', [
+            'env_spec',
+            'observation',
+            'action',
+            'reward',
+            'next_observation',
+            'terminal',
+            'env_info',
+            'agent_info',
+        ])):
+    # pylint: disable=missing-return-doc, missing-return-type-doc, missing-param-doc, missing-type-doc  # noqa: E501
+    r"""A tuple representing a batch of whole trajectories.
+
+    A :class:`Sample` represents single sample sampled from the environment.
+
+    Attributes:
+        env_spec (garage.envs.EnvSpec): Specification for the environment from
+            which this data was sampled.
+        observation (float): Observation at the current time step in the
+            environment.
+        action (float): Action sampled from the policy given the observation
+            at the current time step.
+        reward (float): The reward for taking the action at the current state.
+        terminals (boolean): A boolean representing the termination signal for
```

Comment:
does this data structure need to represent the difference between a true terminal and a timeout terminal? are those treated differently in off-policy algorithms? does it even need to know if a sample is a timeout terminal?